### PR TITLE
Generalize private keys in tsh (PIV integration)

### PR DESCRIPTION
**Pull request by Joerger**
Part 1 for implementing https://github.com/gravitational/teleport/pull/14774 

With PIV integration, we will only have access to a private key interface (`crypto.Signer`, `crypto.PublicKey`). Currently Teleport Client's load private keys from PEM encoded keys rather than their crypto interfaces.

To make room for PIV integration, we need to first change the way we handle private keys to remove our reliance on raw private key PEM.

Changes:
 - Add new `api/utils/keys` package to improve the way we interface with private keys
   - Add new `keys.PrivateKey` interface which implements `crypto.Signer` and provides helper methods for handling tls and ssh certificates.
   - Add `keys.RSAPrivateKey` which wraps `rsa.PrivateKey` to implement `keys.PrivateKey`
   - Add `keys.ParsePrivateKey` to introduce custom private key parsing logic for non-standard keys (like [PIV keys](see https://github.com/gravitational/teleport/pull/15335))
   - Add `keys.[Load]X509KeyPair` to replace calls to `tls.[Load]X509KeyPair` with custom parsing logic above
 - Update native and keygen packages to return `keys.PrivateKey` instead of PEM key
 - Replace key PEM with `keys.PrivateKey` interface in client packages (not `lib/service`, `lib/auth`, etc.)

Minor Changes:
 - Remove unused `agent.AddedKey` return from `LoadKey`
 - Simplify `lib/utils/sshutils` and remove unused code paths
 - Replace `keyagent.certsForCluster("")` with `keyagent.signers`, which gets signers using all certificates in `~/.tsh/keys/proxy/ssh-user/` instead of getting them from the local ssh agent. This is necessary since PIV integration will not support agent keys initially.

Note: I left a couple todos to fix [Kubernetes Integration for non-rsa keys](https://github.com/gravitational/teleport/blob/36e8775c02194c88468f9ee33b02ddb747ad25bc/rfd/0080-yubikey-piv.md#kubernetes-and-mongodb-support). This will be fixed in a follow up PR after the initial implementation of PIV integration, which is focused on SSH access.

### CODE REVIEW

File: api/utils/keypaths/keypaths.go [hunk 1]
```diff
@@ -40,8 +40,6 @@ const (
 	fileExtSSHCert = "-cert.pub"
 	// fileExtPPK is the suffix/extension of a file where an SSH keypair is stored in PuTTY PPK format.
 	fileExtPPK = ".ppk"
-	// fileExtPub is the extension of a file where a public key is stored.
```
Comment (jakule) [inline api/utils/keypaths/keypaths.go:43]: Why is this removed?

File: api/utils/keys/privatekey.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	"github.com/gravitational/trace"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
+)
+
+const (
+	rsaPrivateKeyType = "RSA PRIVATE KEY"
+)
+
+// PrivateKey implements crypto.PrivateKey and crypto.Signer, with additional helper methods
```
Comment (jakule) [inline api/utils/keys/privatekey.go:36]: If this interface implements `crypto.PrivateKey` and `crypto.Signer` then why on `crypto.Signer` is embedded?

File: api/utils/keys/privatekey_test.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+		keyPEM      []byte
+		assertError require.ErrorAssertionFunc
+		expectType  interface{}
+	}{
+		{
+			desc:   "invalid PEM",
+			keyPEM: []byte(`non-pem data`),
+			assertError: func(t require.TestingT, err error, i ...interface{}) {
+				require.Error(t, err)
+				require.True(t, trace.IsBadParameter(err))
```
Comment (jakule) [inline api/utils/keys/privatekey_test.go:40]: nit: Technically, the first assertion is not needed as `nil` is not a bad parameter error. I'm not sure tho, if the error message will be clear enough if you remove it.

**Review by jakule — COMMENTED** [review]
FIrst pass

File: integration/helpers.go [hunk 4]
```diff
@@ -121,27 +121,18 @@ func externalSSHCommand(o commandOptions) (*exec.Cmd, error) {
...
-	}
-	agentKey := agent.AddedKey{
-		PrivateKey:       privateKey,
-		Certificate:      publicKey.(*ssh.Certificate),
-		Comment:          "",
-		LifetimeSecs:     0,
-		ConfirmBeforeUse: false,
-	}
+
+	agentKey := agentKeys[0]
```
[the remark this answers is not in the document]
Comment (Joerger) [reply integration/helpers.go:134]: `key.AsAgentKeys` will only return an empty array if `err != nil`

File: integration/helpers.go [hunk 5]
```diff
@@ -121,27 +121,18 @@ func externalSSHCommand(o commandOptions) (*exec.Cmd, error) {
...
+	agentKey.Comment = ""
```
[the remark this answers is not in the document]
Comment (Joerger) [reply integration/helpers.go:135]: I thought it might overwrite the user's actual ssh-agent, but this isn't the case, so I'll remove the line.

File: api/utils/keys/privatekey.go (hunk 2 above)
Comment (Joerger) [reply api/utils/keys/privatekey.go:36]: `crypto.PrivateKey` is just an `any` type, and has to be type asserted to be used. Here's the godoc for `crypto.PrivateKey`:

```
PrivateKey represents a private key using an unspecified algorithm.

Although this type is an empty interface for backwards compatibility reasons, all private key types in the standard library implement the following interface

interface{
    Public() crypto.PublicKey
    Equal(x crypto.PrivateKey) bool
}
as well as purpose-specific interfaces such as Signer and Decrypter, which can be used for increased type safety within applications.
```

File: api/utils/keypaths/keypaths.go (hunk 1 above)
Comment (Joerger) [reply api/utils/keypaths/keypaths.go:43]: It's no longer being used, since (currently) this PR removes the public key from the `~/.tsh` key store. We can retrieve the public key directly from the private key instead, or from a signed certificate.

To be clear, I removed it because once I refactored the private key usage, it became clear that this file wasn't actually being used for anything. When we would parse the private key, the public key would just be ignored. This public key would only be used directly if we didn't want to parse the private key to get it, but now we are always parsing the private key up front.

If you'd like, I can add the public key back, which we can use to verify that the private key matches the public key on file, but ultimately it won't do much since we can just check that the private key's public key matches the public key in the certificates.

File: api/utils/keys/privatekey_test.go (hunk 3 above)
Comment (Joerger) [reply api/utils/keys/privatekey_test.go:40]: Yep, I usually do this to get a better error message. Alternatively, we could do `require.True(t, trace.IsBadParameter(err), "expected trace.BadParameter, got %T", err)`

File: tool/tsh/tsh.go [hunk 6]
```diff
@@ -2907,24 +2903,27 @@ func makeClientForProxy(cf *CLIConf, proxy string, useProfileLogin bool) (*clien
...
-			if err := c.Agent.Add(k); err != nil {
-				return nil, trace.Wrap(err)
+		if err == nil {
+			for _, k := range agentKeys {
+				if err := c.Agent.Add(k); err != nil {
+					return nil, trace.Wrap(err)
+				}
 			}
+		} else if trace.IsNotImplemented(err) {
+			log.WithError(err).Infof("Failed to add key to agent keys.")
```
[the remark this answers is not in the document]
Comment (Joerger) [reply tool/tsh/tsh.go:2924]: I'll add a comment, but this is not a fatal error. It only breaks Agent forwarding and related integrations, as covered in the RFD.

File: api/utils/keys/privatekey.go [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+	ecPrivateKeyType    = "EC PRIVATE KEY"
+)
+
+// PrivateKey implements crypto.PrivateKey and crypto.Signer, with additional helper methods
+// for performing TLS/SSH handshakes, storing agent keys, and storing private key data.
+type PrivateKey interface {
+	crypto.Signer
+
+	// Equal returns whether the given key is equal to this key
+	Equal(x crypto.PrivateKey) bool
```
Comment (espadolini) [inline api/utils/keys/privatekey.go:49]: It seems like we can do `Equal(x PrivateKey) bool` here (with the appropriate changes), it might be worth it for the added type safety.

File: api/utils/keys/ecdsa.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+// ECDSAPrivateKey is an ecdsa.PrivateKey with additional methods.
+type ECDSAPrivateKey struct {
+	*ecdsa.PrivateKey
+	privateKeyDER []byte
+	sshPub        ssh.PublicKey
+}
+
+// newECDSAPrivateKey creates a new ECDSAPrivateKey from a ecdsa.PrivateKey.
+func newECDSAPrivateKey(priv *ecdsa.PrivateKey) (*ECDSAPrivateKey, error) {
+	keyDER, err := x509.MarshalECPrivateKey(priv)
```
Comment (espadolini) [inline api/utils/keys/ecdsa.go:39]: Any specific reason why we're going for SEC 1 and PKCS#1 here and in rsa.go, rather than always using PKCS#8?

File: api/utils/keys/privatekey.go (hunk 7 above)
Comment (espadolini) [reply api/utils/keys/privatekey.go:49]: It's not like we're getting real interoperability with `crypto.PrivateKey` anyway, all the `crypto` functions will do type assertions and return an error if it's not one of the three builtin ones.

Comment (Joerger) [reply api/utils/keys/privatekey.go]: Good point, I think we can just remove `Equal` entirely. We only need to implement crypto.Signer.

File: api/utils/keys/ecdsa.go (hunk 8 above)
Comment (Joerger) [reply api/utils/keys/ecdsa.go]: No reason other than maintaining old behavior. But I think you're right, we might as well upgrade to PKCS#8 here.

**Review by jakule — APPROVED** [review]

**Review by rosstimothy — COMMENTED** [review]
Not to totally derail this PR but did you consider a generic approach? The `PrivateKey` implementations all look identical. Something like the following seems like it would simplify some the `keys` package quite a bit, albeit doing so would limit where this could be backported to, so if this is intended to exist prior to v10 it's a bit of a moot point. 

```go
type Key interface {
	crypto.Signer
	crypto.PrivateKey
}

type PrivateKey[T Key] struct {
	key           T
	privateKeyDER []byte
	sshPub        ssh.PublicKey
}

func (p PrivateKey[T]) Public() crypto.PublicKey {
	return p.key.Public()
}

func (p PrivateKey[T]) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
	return p.key.Sign(rand, digest, opts)
}

// PrivateKeyPEM returns the PEM encoded ECDSA private key.
func (p PrivateKey[T]) PrivateKeyPEM() []byte {
	return pem.EncodeToMemory(&pem.Block{
		Type:    pkcs8PrivateKeyType,
		Headers: nil,
		Bytes:   p.privateKeyDER,
	})
}

// SSHPublicKey returns the ssh.PublicKey representation of the public key.
func (p PrivateKey[T]) SSHPublicKey() ssh.PublicKey {
	return p.sshPub
}

// TLSCertificate parses the given TLS certificate paired with the private key
// to return a tls.Certificate, ready to be used in a TLS handshake.
func (p PrivateKey[T]) TLSCertificate(certRaw []byte) (tls.Certificate, error) {
	cert, err := tls.X509KeyPair(certRaw, p.PrivateKeyPEM())
	return cert, trace.Wrap(err)
}

```

**Comment by Joerger** [comment]
@rosstimothy Thanks for your suggestion, the only issue is that in order to support non-standard private keys, `TLSCertificate` and `PrivateKeyPEM` will need to be variable between private keys (take a look at https://github.com/gravitational/teleport/pull/15335 for context). I went ahead and implemented something similar to your suggestion, let me know what you think.

**Comment by Joerger** [comment]
This PR is currently blocked this flaky test issue: https://github.com/gravitational/teleport/issues/15816

This PR makes that test run ~250ms slower, which is enough to cause the error more frequently. This must be related to how the Proxy Service handles client keys, but I'm not sure what specifically is causing this.

**Review by rosstimothy — APPROVED** [review]
